util/defer.h

```diff
@@ -49,4 +49,22 @@ class Defer final {
   std::function<void()> fn_;
 };
 
+// An RAII utility object that saves the current value of an object so that
+// it can be overwritten, and restores it to the saved value when the
+// SaveAndRestore object goes out of scope.
+template <typename T>
+class SaveAndRestore {
+ public:
+  explicit SaveAndRestore(T* obj) : obj_(obj), saved_(*obj) {}
```

Comment:
Being paranoid, but I think it is safe to make sure `obj` is not nullptr.

Reply:
By the time we can write `assert(obj)` we have already used it. I can put it in a comment though